fs/backgroundfetcher/background_fetcher.go

```diff
@@ -0,0 +1,167 @@
+}
+
+type defaultPauser struct{}
+
+func (p defaultPauser) pause(d time.Duration) {
+	time.Sleep(d)
+}
+
+// A backgroundFetcher is responsible for fetching spans from layers
+// in the background.
+type BackgroundFetcher struct {
+	silencePeriod time.Duration
+	fetchPeriod   time.Duration
+	maxQueueSize  int
+
+	rateLimiter *rate.Limiter
+
+	bfPauser pauser
+
+	// All span managers are added to the channel and picked up in Run().
+	// If a span manager is still able to fetch, it is reinserted into the chanel.
+	workQueue chan Resolver
+	closeChan chan struct{}
+	pauseChan chan struct{}
+}
+
+func NewBackgroundFetcher(opts ...Option) (*BackgroundFetcher, error) {
+	bf := new(BackgroundFetcher)
+	for _, o := range opts {
+		if err := o(bf); err != nil {
+			return nil, err
+		}
+	}
+	// Create a rate-limiter that will fetch every bf.fetchPeriod
+	// with a burst capacity of 1 (i.e., it will never invoke more than 1 bg-fetch
+	// within bf.fetchPeriod)
+	bf.rateLimiter = rate.NewLimiter(rate.Every(bf.fetchPeriod), 1)
+	bf.workQueue = make(chan Resolver, bf.maxQueueSize)
+	bf.closeChan = make(chan struct{})
+	bf.pauseChan = make(chan struct{})
+
+	if bf.bfPauser == nil {
+		bf.bfPauser = defaultPauser{}
+	}
+
+	return bf, nil
+}
+
+// Add a new Resolver to be background fetched from.
+// Sends the resolver through the channel, which will be received in the Run() method.
+func (bf *BackgroundFetcher) Add(resolver Resolver) {
+	bf.workQueue <- resolver
+}
+
+func (bf *BackgroundFetcher) Close() error {
+	bf.closeChan <- struct{}{}
+	return nil
+}
+
+// Sends a signal to pause the background fetcher for silencePeriod on the next iteration.
+func (bf *BackgroundFetcher) Pause() {
+	bf.pauseChan <- struct{}{}
+}
+
+func (bf *BackgroundFetcher) pause(ctx context.Context) {
+	needPause := false
+loop:
+	for {
+		select {
+		// A new image has been mounted. Need to pause the background fetcher
+		case <-bf.pauseChan:
+			needPause = true
+		default:
+			break loop
+		}
+	}
+	if needPause {
+		log.G(ctx).WithField("silencePeriod", bf.silencePeriod).Debug("new image mounted, pausing the background fetcher for silence period")
+		bf.bfPauser.pause(bf.silencePeriod)
+	}
+}
+
+func (bf *BackgroundFetcher) Run(ctx context.Context) error {
+	for {
+		// Pause the background fetcher if necessary.
+		bf.pause(ctx)
+
+		select {
+		case <-bf.closeChan:
+			return nil
+		case <-ctx.Done():
+			return nil
+		case lr := <-bf.workQueue:
+			if lr.Closed() {
+				continue
+			}
+			go func() {
+				more, err := lr.Resolve(ctx)
+				if more {
+					bf.workQueue <- lr
+				} else if err != nil {
+					log.G(ctx).WithError(err).Warn("error trying to resolve layer, removing it from the queue")
+				}
+			}()
+		default:
+		}
+
+		if err := bf.rateLimiter.Wait(ctx); err != nil {
+			log.G(ctx).WithError(err).Warn("background fetcher: error while waiting for rate limiter")
+			return fmt.Errorf("background fetch: error while waiting for rate limiter: %w", err)
```

Comment:
Generally speaking, it is better to do either logging (because the caller cannot catch the error) or returning an error (bacause the caller can catch the error), not both.

Reply:
Yes, will submit a new revision that will address this.